src/Payum/Core/Tests/Storage/CryptoStorageDecoratorTest.php

```diff
@@ -0,0 +1,234 @@
+
+        $storage->identify($model);
+    }
+
+    public function testShouldProxyCallToDecoratedStorageAndPassCypherToModelEncryptOnUpdate()
+    {
+        $model = new CryptedModel();
+
+        $decoratedStorage = $this->createStorageMock();
+        $decoratedStorage
+            ->expects($this->once())
+            ->method('update')
+            ->willReturn($model)
+        ;
+
+        $cypherMock = $this->createCypherMock();
+        $cypherMock
+            ->expects($this->once())
+            ->method('encrypt')
+            ->with('theVal');
+        ;
+        $cypherMock
+            ->expects($this->never())
+            ->method('decrypt')
+        ;
+
+        $storage = new CryptoStorageDecorator($decoratedStorage, $cypherMock);
+
+        $storage->update($model);
+    }
+
+    public function testShouldProxyCallToDecoratedStorageAndPassCypherToModelDecryptOnFind()
+    {
+        $model = new CryptedModel();
+
+        $decoratedStorage = $this->createStorageMock();
+        $decoratedStorage
+            ->expects($this->once())
+            ->method('find')
+            ->willReturn($model)
+        ;
+
+        $cypherMock = $this->createCypherMock();
+        $cypherMock
+            ->expects($this->never())
+            ->method('encrypt')
+        ;
+        $cypherMock
+            ->expects($this->once())
+            ->method('decrypt')
+            ->with('theEncryptedVal');
+        ;
+
+        $storage = new CryptoStorageDecorator($decoratedStorage, $cypherMock);
+
+        $storage->find('anId');
+    }
+
+    public function testShouldProxyCallToDecoratedStorageAndPassCypherToEveryModelDecryptOnFindBy()
+    {
+        $decoratedStorage = $this->createStorageMock();
+        $decoratedStorage
+            ->expects($this->once())
+            ->method('findBy')
+            ->willReturn([new CryptedModel(), new CryptedModel(), new \stdClass()])
+        ;
+
+        $cypherMock = $this->createCypherMock();
+        $cypherMock
+            ->expects($this->never())
+            ->method('encrypt')
+        ;
+        $cypherMock
+            ->expects($this->exactly(2))
+            ->method('decrypt')
+            ->with('theEncryptedVal');
+        ;
+
+        $storage = new CryptoStorageDecorator($decoratedStorage, $cypherMock);
+
+        $storage->findBy([]);
+    }
+
+    /**
+     * @return \PHPUnit_Framework_MockObject_MockObject|StorageInterface
+     */
+    private function createStorageMock()
+    {
+        return $this->getMock(StorageInterface::class);
+    }
+
+    /**
+     * @return \PHPUnit_Framework_MockObject_MockObject|CypherInterface
+     */
+    private function createCypherMock()
+    {
+        return $this->getMock(CypherInterface::class);
+    }
+}
+
+class CryptedModel implements CryptedInterface
+{
+    public function decrypt(CypherInterface $cypher)
+    {
+        $cypher->decrypt('theEncryptedVal');
```

Comment:
Oh, all right, `CypherInterface` and `CryptedInterface` have methods name the same and I just messed it up.